### PR TITLE
feat: Add filtering and sorting on userTag screens (GSoC)

### LINKED ISSUE #2395: ADMIN: Tag screens filter and sort (GSoC)
### Is your feature request related to a problem? Please describe.

Adding filtering and sorting for better user experience.

### Additional context

This is a companion issue of the one in the api: PalisadoesFoundation/talawa-api#2630

### PR DESCRIPTION
**Pull request by meetulr**
### What kind of change does this PR introduce?

Adds filtering and sorting for userTag, refactors and addresses minor issues.

### Issue Number:

Fixes #2395

### Did you add tests for your changes?

Yes

### Demo

https://github.com/user-attachments/assets/5f101e8f-1ea5-4ae9-ac42-4104bb3d8673

### Additional Context

talawa-api issue: PalisadoesFoundation/talawa-api#2630

## Summary by CodeRabbit

- **New Features**
	- Added search functionality for tags based on user input, allowing filtering by name.
	- Introduced sorting capabilities for tags, enabling users to sort by latest or oldest first.
	- Enhanced user interface with prompts for entering tag names in multiple languages.
	- Improved loading and display of tags with breadcrumb navigation for ancestor tags.

- **Bug Fixes**
	- Improved error handling for tag assignments and creation processes.

- **Tests**
	- Expanded test coverage to include scenarios for searching and sorting tags, as well as error handling during tag creation.

- **Documentation**
	- Updated interfaces to include new properties for parent and ancestor tags.

### CODE REVIEW

File: src/GraphQl/Queries/userTagQueries.ts [hunk 1]
```diff
@@ -98,6 +123,7 @@ export const USER_TAGS_MEMBERS_TO_ASSIGN_TO = gql`
     $before: String
     $first: PositiveInt
     $last: PositiveInt
+    $where: UserTagUsersToAssignToWhereInput
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/GraphQl/Queries/userTagQueries.ts:126]: It'll be added if required.

File: src/screens/OrganizationTags/OrganizationTags.test.tsx [hunk 2]
```diff
@@ -160,7 +160,85 @@ describe('Organisation Tags Page', () => {
...
+    });
+    userEvent.click(screen.getByTestId('latest'));
+
+    // reverse the order again
+    await waitFor(() => {
+      expect(screen.getAllByTestId('tagName')[0]).toHaveTextContent(
+        'searchUserTag 1',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/OrganizationTags/OrganizationTags.test.tsx]: It's fine.

File: src/components/AddPeopleToTag/AddPeopleToTag.test.tsx [hunk 3]
```diff
@@ -157,13 +250,15 @@ describe('Organisation Tags Page', () => {
...
+      'addPeopleToTagScrollableDiv',
+    );
 
     const initialMemberDataLength = screen.getAllByTestId('memberName').length;
 
     // Set scroll position to the bottom
-    fireEvent.scroll(scrollableDiv, {
-      target: { scrollY: scrollableDiv.scrollHeight },
+    fireEvent.scroll(addPeopleToTagScrollableDiv, {
+      target: { scrollY: addPeopleToTagScrollableDiv.scrollHeight },
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/components/AddPeopleToTag/AddPeopleToTag.test.tsx]: It's fine.

File: src/screens/SubTags/SubTags.test.tsx [hunk 4]
```diff
@@ -238,6 +236,84 @@ describe('Organisation Tags Page', () => {
...
+    });
+    userEvent.click(screen.getByTestId('latest'));
+
+    // reverse the order again
+    await waitFor(() => {
+      expect(screen.getAllByTestId('tagName')[0]).toHaveTextContent(
+        'searchSubTag 1',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/SubTags/SubTags.test.tsx]: It's fine.

File: src/utils/interfaces.ts [hunk 5]
```diff
@@ -258,16 +263,19 @@ export interface InterfaceQueryOrganizationUserTags {
 export interface InterfaceQueryUserTagChildTags {
   name: string;
   childTags: InterfaceTagNodeData;
+  ancestorTags: {
+    _id: string;
+    name: string;
+  }[];
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/utils/interfaces.ts]: It does have it.

File: src/components/TagActions/TagActions.tsx [hunk 6]
```diff
@@ -237,6 +213,11 @@ const TagActions: React.FC<InterfaceTagActionsProps> = ({
   ): Promise<void> => {
     e.preventDefault();
 
+    if (!selectedTags.length) {
+      toast.error(t('noTagSelected'));
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/components/TagActions/TagActions.tsx]: disabling it won't look good. Proper handling is being done.

File: src/components/TagActions/TagActions.tsx [hunk 7]
```diff
@@ -353,17 +348,36 @@
...
+                              <span
+                                key={ancestorTag._id}
+                                className="ms-2 my-0"
+                                data-testid="ancestorTagsBreadCrumbs"
+                              >
+                                {ancestorTag.name}
+                                <i className="ms-2 fa fa-caret-right" />
+                              </span>
+                            ))}
+                            <>{')'}</>
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/components/TagActions/TagActions.tsx]: It looks fine.

File: src/screens/OrganizationTags/OrganizationTags.tsx [hunk 8]
```diff
@@ -272,21 +302,16 @@ function OrganizationTags(): JSX.Element {
...
+              <i className="fa fa-search position-absolute text-body-tertiary end-0 top-50 translate-middle" />
               <Form.Control
                 type="text"
                 id="tagName"
                 className="bg-white"
-                placeholder={tCommon('search')}
+                placeholder={tCommon('searchByName')}
                 data-testid="searchByName"
+                onChange={(e) => setTagSearchName(e.target.value.trim())}
                 autoComplete="off"
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/OrganizationTags/OrganizationTags.tsx]: something for later.

File: src/screens/ManageTag/ManageTag.test.tsx [hunk 9]
```diff
@@ -339,6 +324,84 @@ describe('Manage Tag Page', () => {
...
+    });
+    userEvent.click(screen.getByTestId('latest'));
+
+    // reverse the order again
+    await waitFor(() => {
+      expect(screen.getAllByTestId('memberName')[0]).toHaveTextContent(
+        'assigned user1',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/ManageTag/ManageTag.test.tsx]: It'll do.

File: src/screens/SubTags/SubTags.tsx [hunk 10]
```diff
@@ -296,21 +282,16 @@ function SubTags(): JSX.Element {
...
                 type="text"
                 id="tagName"
                 className="bg-white"
-                placeholder={tCommon('search')}
+                placeholder={tCommon('searchByName')}
+                onChange={(e) => setTagSearchName(e.target.value.trim())}
                 data-testid="searchByName"
                 autoComplete="off"
-                required
               />
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/SubTags/SubTags.tsx]: something for later.

File: src/screens/SubTags/SubTags.tsx [hunk 11]
```diff
@@ -71,6 +74,8 @@
     variables: {
       id: parentTagId,
       first: TAGS_QUERY_DATA_CHUNK_SIZE,
+      where: { name: { starts_with: tagSearchName } },
+      sortedBy: { id: tagSortOrder },
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/SubTags/SubTags.tsx]: It's according the the types defined in the api. More filtering criteria can be added later.

File: src/screens/ManageTag/ManageTag.tsx [hunk 12]
```diff
@@ -277,15 +259,23 @@ function ManageTag(): JSX.Element {
...
+
+  // get the ancestorTags array and push the current tag in it
+  // used for the tag breadcrumbs
+  const orgUserTagAncestors = [
+    ...(userTagAssignedMembersData?.getAssignedUsers.ancestorTags ?? []),
+    {
+      _id: currentTagId,
+      name: currentTagName,
+    },
+  ];
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/ManageTag/ManageTag.tsx]: It's not an empty string.

File: src/screens/ManageTag/ManageTag.tsx [hunk 13]
```diff
@@ -146,26 +150,14 @@
...
-    },
-  });
+  useEffect(() => {
+    const [firstName, ...lastNameParts] = assignedMemberSearchInput
+      .trim()
+      .split(/\s+/);
+    const lastName = lastNameParts.join(' '); // Joins everything after the first word
+    setAssignedMemberSearchFirstName(firstName);
+    setAssignedMemberSearchLastName(lastName);
+  }, [assignedMemberSearchInput]);
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/ManageTag/ManageTag.tsx]: It's being handled.

File: src/screens/ManageTag/ManageTag.tsx [hunk 14]
```diff
@@ -107,7 +105,13 @@
     variables: {
       id: currentTagId,
       first: TAGS_QUERY_DATA_CHUNK_SIZE,
+      where: {
+        firstName: { starts_with: assignedMemberSearchFirstName },
+        lastName: { starts_with: assignedMemberSearchLastName },
+      },
+      sortedBy: { id: assignedMemberSortOrder },
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/ManageTag/ManageTag.tsx]: It is designed to match all entries if filters are empty strings.

File: src/components/AddPeopleToTag/AddPeopleToTag.test.tsx [hunk 15]
```diff
@@ -147,6 +174,72 @@ describe('Organisation Tags Page', () => {
...
+        'first userToAssignTo',
+      );
+    });
+
+    await waitFor(() => {
+      expect(screen.getAllByTestId('memberName')[1]).toHaveTextContent(
+        'second userToAssignTo',
+      );
+    });
+  });
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/components/AddPeopleToTag/AddPeopleToTag.test.tsx:241]: It's fine, it's more readable this way.

**Review by rishav-jha-mech — APPROVED** [review]

File: src/screens/OrganizationTags/OrganizationTags.tsx [hunk 16]
```diff
@@ -109,12 +121,21 @@ function OrganizationTags(): JSX.Element {
...
     useMutation(CREATE_USER_TAG);
 
   const createTag = async (e: ChangeEvent<HTMLFormElement>): Promise<void> => {
     e.preventDefault();
 
+    if (!tagName.trim()) {
+      toast.error(t('enterTagName'));
+      return;
+    }
+
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/OrganizationTags/OrganizationTags.tsx:138]: Something for later.

File: src/screens/OrganizationTags/OrganizationTags.tsx [hunk 17]
```diff
@@ -322,61 +355,66 @@
...
+                    rowHeight={65}
+                    rows={userTagsList?.map((userTag, index) => ({
+                      id: index + 1,
+                      ...userTag,
+                    }))}
+                    columns={columns}
+                    isRowSelectable={() => false}
+                  />
+                </InfiniteScroll>
+              </div>
```
[the remark this answers is not in the document]
Comment (meetulr) [reply src/screens/OrganizationTags/OrganizationTags.tsx:415]: Something for later, would be required in all screens.

**Pull request merged**